### PR TITLE
[#14] Improve Has instances experience

### LINKED ISSUE #14: Improve Has instances experience
See https://github.com/Holmusk/cake-slayer/pull/12#discussion_r327008670

### PR DESCRIPTION
**Pull request by vrom911**
Resolves #14

### CODE REVIEW

File: src/CakeSlayer/Has.hs [hunk 1]
```diff
@@ -56,3 +65,69 @@ class Has field env where
...
+__instance__ 'HasField' "envInt" Env 'Int' __where__
+    'getField' :: Env -> 'Int'
+    'getField' = envInt
+
+__instance__ 'HasField' "envString" Env 'String' __where__
+    'getField' :: Env -> 'String'
+    'getField' = envString
+@
+
+You can have now see that using @'getField' \@\"envInt\"@ is equivalent to @envInt@.
```
Comment (chshersh) [inline src/CakeSlayer/Has.hs]:
```suggestion
You can now see that using @'getField' \@\"envInt\"@ is equivalent to @envInt@.
```

File: src/CakeSlayer/Has.hs [hunk 2]
```diff
@@ -56,3 +65,69 @@ class Has field env where
...
+@
+
+It should be possible to write it using `DerivingVia` extension:
+
+@
+data Env = Env
+    { envInt :: !Int
+    , envString :: !String
+    } deriving (Has Int) via Field "envInt" Int Env
+      deriving (Has String) via Field "envString" String Env
```
Comment (chshersh) [inline src/CakeSlayer/Has.hs]: Is it good idea to omit `()` here and also use bold for `deriving` and `via` words? :thinking: 
```suggestion
      deriving Has String via Field "envString" String Env
```

File: src/CakeSlayer/Has.hs [hunk 3]
```diff
@@ -56,3 +65,69 @@ class Has field env where
...
+@
+
+__ NOTE:__ This only works starting with @GHC-8.6@.
+-}
+newtype Field (s :: Symbol) field env = Field
+    { unField :: env
+    }
+
+instance forall s f env . (HasField s env f) => Has f (Field s f env) where
+    obtain = getField @s . unField
```
Comment (chshersh) [inline src/CakeSlayer/Has.hs]: Let's add type signature to `obtain` in this instance. Just to make it easier to understand the code...

File: src/CakeSlayer/Has.hs (hunk 2 above)
Comment (vrom911) [reply src/CakeSlayer/Has.hs]: It doesn't parse without parenthesis...

Comment (chshersh) [reply src/CakeSlayer/Has.hs]: Ah, I didn't know that!

File: src/CakeSlayer/Has.hs [hunk 4]
```diff
@@ -104,22 +105,24 @@ __data__ Env = Env
     }
 
 __instance__ 'Has' 'Int' Env __where__
+    'obtain' :: Env -> 'Int'
     'obtain' = envInt
-    {-# INLINE 'obtain' #-}
+    \{\-\# INLINE 'obtain' \#\-\}
```
Comment (chshersh) [inline src/CakeSlayer/Has.hs]: Lol. Nice Haddock :+1: Sometimes it's weird...

**Pull request merged**